cores/arduino/printk.cpp

```diff
@@ -0,0 +1,230 @@
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdarg.h>
+#include "UARTClass.h"
+
+extern "C" void printk(const char *fmt, va_list args);
+int __vsnprintf(char *dest, int size, const char *fmt, va_list args)
+{
+	int might_format = 0;
+	int len = 0;
+	char lastchar = 0;
+	bool binary_format = false;
+
+	if (!dest || !size)
+		return 0;
+
+	while (*fmt && len < size) {
+		if (!might_format) {
+			if (*fmt == '\n' && lastchar != '\r') {
+				if (len < size) {
+					lastchar = *dest++ = '\r', len++;
+					continue;
+				} else
+					break;
+			} else if (*fmt != '%') {
+				if (len < size)
+					lastchar = *dest++ = *fmt, len++;
+				else
+					break;
+			} else
+				might_format = 1;
+		} else {
+			if (*fmt == '%') {
+				if (len < size)
+					*dest++ = '%', len++;
+				else
+					break;
+				might_format = 0;
+			} else {
+				switch (*fmt) {
+				case '0':
+				case '1':
+					might_format |= 2;
+					goto still_format;
+					break;
+				case '2':
+					might_format |= 4;
+					goto still_format;
+					break;
+				case '4':
+					might_format |= 8;
+					goto still_format;
+					break;
+				case 'b':
+					binary_format = true;
+					goto still_format;
+					break;
+				case 'd':
+				case 'i':
+				case 'u':
+					if (!binary_format) {
+						unsigned long num =
+							va_arg(args,
+							       unsigned long);
+						unsigned long pos = 999999999;
+						unsigned long remainder = num;
+						int found_largest_digit = 0;
+
+						if (*fmt != 'u'
+						    && (num & (1 << 31))) {
+							if (len < size)
+								*dest++ =
+									'-',
+								len++;
+							num = (~num) + 1;
+							remainder = num;
+						}
+						while (pos >= 9) {
+							if (found_largest_digit
+							    || remainder >
+							    pos) {
+								found_largest_digit
+									= 1;
+								if (len < size)
+									*dest++
+										=
```

Comment:
Looks like the formatting got a little messed up for the next 10 lines here, can you clean it up please?